index.ts

```diff
@@ -1 +1,163 @@
-console.log('hello repo');
+enum OperationType {
+  AddItemQty = 'AddItemQty',
+  AddProductStock = 'AddProductStock',
+  AddProductUnitPrice = 'AddProductUnitPrice',
+  SetSellerInfo = 'SetSellerInfo',
+  SetBuyerInfo = 'SetBuyerInfo'
+}
+
+enum Actor {
+  Buyer = 'Buyer',
+  Seller = 'Seller'
+}
+
+const OperationsAllowed = {
+  [Actor.Seller]: [
+    OperationType.AddProductStock,
+    OperationType.AddProductUnitPrice,
+    OperationType.SetSellerInfo
+  ],
+  [Actor.Buyer]: [
+    OperationType.AddItemQty,
+    OperationType.SetBuyerInfo
+  ]
+}
+
+class ItemsOperation {
+  type: OperationType
+  productCode: string
+  amount: number
+}
+
+class InfoOperation {
+  type: OperationType
+  version: number
+  info: any
+}
+
+class DraftInvoice {
+  actorInfo: {
+    [actor: string]: { // actually, [actor: Actor]
+      info: any
+      version: number
+    }
+  }
+  invoiceId: string
+  items: {
+   [productCode: string]: {
+      ask: number
+      stock: number
+      unitPrice: number
+    }
+  }
+
+  constructor(invoiceId: string, sellerInfo: any, buyerInfo: any) {
+    this.invoiceId = invoiceId;
+    this.actorInfo = {
+      [Actor.Seller]: { info: sellerInfo, version: 0 },
+      [Actor.Buyer]: { info: buyerInfo, version: 0 }
+    }
+    this.items = {};
+    console.log(`Draft invoice ${invoiceId} created, seller and buyer:`, sellerInfo, buyerInfo);
+  }
+
+  isInfoOperation(op: ItemsOperation | InfoOperation): boolean {
+    // FIXME: why doesn't this work:
+    // return op instanceof ItemsOperation;
+    return ([OperationType.SetBuyerInfo, OperationType.SetSellerInfo].indexOf(op.type) !== -1);
+  }
+  
+  applyOperation (op: ItemsOperation | InfoOperation): void {
+    if (this.isInfoOperation(op)) {
+      this.applyInfoOperation(op as InfoOperation)
+    } else {
+      this.applyItemsOperation(op as ItemsOperation)
+    }
+  }
+
+  applyInfoOperation(op: InfoOperation) {
+    let actor: Actor
+    if (op.type === OperationType.SetSellerInfo) {
+      actor = Actor.Seller;
+    } else if (op.type === OperationType.SetBuyerInfo) {
+      actor = Actor.Buyer;
+    } else {
+      throw new Error(`Unknown info operation type ${op.type}`)
+    }
+    if (op.version === this.actorInfo[actor].version) {
+      throw new Error('Conflict! Same version number used twice');
+    }
+    if (op.version < this.actorInfo[actor].version) {
+      return; // old news, ignore
+    }
+    console.log(`Setting ${(actor === Actor.Buyer ? 'buyer' : 'seller')} info version ${op.version}:`, op.info);
+    this.actorInfo[actor] = { info: op.info, version: op.version };
+  }
+
+  applyItemsOperation(op: ItemsOperation) {
+    if (!this.items[op.productCode]) {
+      this.items[op.productCode] = {
+        ask: 0,
+        stock: 0,
+        unitPrice: 0
+      }
+    }
```

Comment:
So here is a stealthy grow-only set of product-code-items. Its one operation (add product) is hidden in adding some amount. But it's definitely conflict-free! Things would get more exciting if we wanted to remove from this set too.